src/server/server.py

```diff
@@ -99,7 +113,118 @@
             print('Option reading error (ssl).')
         exit(1)
 
+# Cluster mode is used for revocation
+try:
+    SERVER_OPTS['cluster'] = CONFIG.get('main', 'cluster').split(',')
+except NoOptionError:
+    # Standalone mode
+    PROTO = 'http'
+    if SERVER_OPTS['ssl']:
+        PROTO = 'https'
+    SERVER_OPTS['cluster'] = ['%s://localhost:%s' % (PROTO, SERVER_OPTS['port'])]
+
+try:
+    SERVER_OPTS['clustersecret'] = CONFIG.get('main', 'clustersecret')
+except NoOptionError:
+    # Standalone mode
+    SERVER_OPTS['clustersecret'] = randomString(32)
+
+
 # INDEPENDANT FUNCTIONS
+def get(url):
+    """
+    Rebuilt GET function for CASSH purpose
+    """
+    session = Session()
+    try:
+        req = session.get(url,
+                          headers=REQ_HEADERS,
+                          timeout=REQ_TIMEOUT,
+                          stream=True)
+    except ConnectionError:
+        print('Connection error : %s' % url)
+        req = None
+    return req
+
+def post(url, payload):
+    """
+    Rebuilt POST function for CASSH purpose
+    """
+    session = Session()
+    try:
+        req = session.post(url,
+                           data=payload,
+                           headers=REQ_HEADERS,
+                           timeout=REQ_TIMEOUT)
+    except ConnectionError:
+        print('Connection error : %s' % url)
+        req = None
+    return req
+
+def randomString(stringLength=10):
+    """Generate a random string of fixed length """
+    letters = ascii_lowercase
+    return ''.join(choice(letters) for i in range(stringLength))
+
+def cluster_alived():
+    """
+    This function returns a subset of pingeable node
+    """
+    alive_nodes = list()
+    dead_nodes = list()
+    for node in SERVER_OPTS['cluster']:
+        req = get("%s/ping" % node)
+        if req is not None and req.text == 'pong':
+            alive_nodes.append(node)
+        else:
+            dead_nodes.append(node)
+    return alive_nodes, dead_nodes
+
+def cluster_last_krl():
+    """
+    This functions download the biggest KRL of the cluster.
+    It's not perfect...
+    """
+    subset, _ = cluster_alived()
+    cluster_id = 0
+    for node in subset:
+        req = get("%s/krl" % node)
+        with open('/tmp/%s.krl' % cluster_id, 'wb') as cluster_file:
+            for chunk in req.iter_content(chunk_size=1024):
+                if chunk: # filter out keep-alive new chunks
+                    cluster_file.write(chunk)
+        cluster_id += 1
+
+    top_size = stat(SERVER_OPTS['krl']).st_size
+    top_cluster_id = 'local'
+    for i in range(cluster_id):
+        if stat('/tmp/%s.krl' % i).st_size >= top_size:
+            top_cluster_id = i
+
+    print('%s KRL is the best' % top_cluster_id)
+
+    if top_cluster_id != 'local':
+        move('/tmp/%s.krl' % top_cluster_id, SERVER_OPTS['krl'])
+    else:
+        cluster_updatekrl(None, update_only=True)
+
+    return True
+
+
+def cluster_updatekrl(username, update_only=False):
```

Comment:
If I refer to l183, you have named the function ` cluster_last_krl`,
should this one be named ` cluster_update_krl` rather than `cluster_updatekrl` ?

Reply:
The same, because I need to make more changes :
https://github.com/nbeguier/cassh/issues/67